### PR TITLE
[fix] OSS dict load/save fix - better fix than 383 and unit test

### LINKED ISSUE #380: [OSS] When loading a state, the insertion order for the params may not match the params_groups
## 🐛 Bug
(FB Only f250383189) Guess is that the issue comes from the parameters not being inserted in the right order in the sharded optimizer state. [See](https://github.com/facebookresearch/fairscale/blob/master/fairscale/optim/oss.py#L403)

Would need fixing + repro in a unit test

### PR DESCRIPTION
**Pull request by blefaudeux**
# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
- [x] Did you read the [contributor guideline](https://github.com/facebookresearch/fairscale/blob/master/CONTRIBUTING.md)?
- [ ] Did you make sure to update the docs?
- [x] Did you write any new necessary tests? Catch the case which was broken before, and fix it

## What does this PR do?
Fixes #380. Better take than #383 because fixing another issue which was not caught (383 was not enough), and reproducing the issue in an updated unit test so that this does not happen again. Thanks again @zhengwy888

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: tests/optim/test_oss.py [hunk 1]
```diff
@@ -859,10 +861,16 @@ def closure_sharded(input_tensor=input_tensor):
         sharded_optim_state_dict = sync_object_ranks(sharded_optim_state_dict, RECIPIENT_RANK, device)
 
         # - cross load the states
+        # run one step and check that the models are still the same
```
Comment (blefaudeux) [inline tests/optim/test_oss.py:864]: this modified unit test does break on the old version

File: fairscale/optim/oss.py [hunk 2]
```diff
@@ -379,6 +379,8 @@ def state_dict(self) -> Dict[str, Any]:
                         global_id = self.param_to_index[local_index_to_param_id[local_param_index]]
                         state_dict["state"][global_id] = s["state"][local_param_index]
 
+        # Make sure that the parameters are sorted in the state, as expected
+        state_dict["state"] = dict(sorted(state_dict["state"].items()))
```
Comment (blefaudeux) [inline fairscale/optim/oss.py:383]: the state dict returned was sorted properly under the "param_groups" key, but not under the "state" field. Pytorch just uses the ordering from the "param_groups" key, and I was just testing the loading OSS-> Pytorch and vice versa, so t his was not caught

File: fairscale/optim/oss.py [hunk 3]
```diff
@@ -391,23 +393,16 @@ def load_state_dict(self, state_dict: Dict[str, Any]) -> None:
...
             # Populate the sharded optimizer state on the fly
             if self.param_to_rank[param] != self.rank:
                 state_dict["state"][key] = None
-
-            if key in self.index_to_param:
-                param = self.index_to_param[i_param]
-
-                # Only add this state to the sharded optimizer if it owns this param
-                for pg in self.optim.param_groups:
-                    if id(param) in [id(p) for p in pg["params"]]:
```
Comment (blefaudeux) [inline fairscale/optim/oss.py:407]: this second check could mask an issue, we just checked above that this rank owns this param, so this is not needed (and potentially risky)

File: tests/optim/test_oss.py [hunk 4]
```diff
@@ -832,8 +834,8 @@ def closure_sharded(input_tensor=input_tensor):
             loss_sharded_optim = cast(torch.Tensor, sharded_optimizer.step(closure=closure_sharded))
 
             assert torch.allclose(
-                loss_ddp, loss_sharded_optim
-            ), f"Losses differ in between Pytorch optim and OSS\nworld size {world_size}"
+                loss_ddp, loss_sharded_optim, rtol=1e-3
```
Comment (blefaudeux) [inline tests/optim/test_oss.py:837]: the rtol change is only needed on pytorch 1.5 unfortunately, without that on a two gpu machine the difference becomes 
26.404895782470703 vs 26.404342651367188 (which I assume is due to a different casting and not structurally wrong) and this asserts

**Comment by blefaudeux** [comment]
sorry @min-xu-ai for the revert of the previous one, I just thought this was cleaner and there was one fix left in the cold

File: tests/optim/test_oss.py (hunk 4 above)
Comment (min-xu-ai) [reply tests/optim/test_oss.py:837]: I think it is worth documenting the reason for 1e-3.

**Review by min-xu-ai — APPROVED** [review]
yeah, this seems to be much nicer.

File: fairscale/optim/oss.py (hunk 2 above)
Comment (zhengwy888) [reply fairscale/optim/oss.py:383]: I didn't know python's dictionary is ordered, so I just looked it up. turns out this has been enabled since python 3.5, good to know! https://stackoverflow.com/questions/39980323/are-dictionaries-ordered-in-python-3-6
